In [ ]:
import pandas as pd

In [ ]:
df=pd.read_csv('https://raw.githubusercontent.com/anvarnarz/praktikum_datasets/main/uzbekistan.csv', index_col=0)
df

Maydoni (kv.km)    Aholisi
Boʻlinishi                                               
Andijon viloyati                        4200.0  1899000.0
Buxoro viloyati                        39400.0  1384700.0
Fargʻona viloyati                       6800.0  2597000.0
Jizzax viloyati                        20500.0   910500.0
Xorazm viloyati                         6300.0  1200000.0
Namangan viloyati                       7900.0  1862000.0
Navoiy viloyati                       110800.0   767500.0
Qashqadaryo viloyati                   28400.0  2029000.0
Qoraqalpogʻiston Respublikasi         160000.0  1200000.0
Samarqand viloyati                     16400.0  2322000.0
Sirdaryo viloyati                       5100.0   648100.0
Surxondaryo viloyati                   20800.0  1676000.0
Toshkent viloyati                      15300.0  4450000.0

In [ ]:
df.sum()

Maydoni (kv.km)      441900.0
Aholisi            22945800.0
dtype: float64

In [ ]:
df2=pd.read_csv("https://raw.githubusercontent.com/anvarnarz/praktikum_datasets/main/gm_sales.csv", index_col=0)
df2

2020   2021
Labo     4521   7542
Damas   30889  33682
Spark   16370   6485
Nexia   41663  14442
Cobalt  44901  36317
Gentra  30368  26266

In [ ]:
df2.sum()

2020    168712
2021    124734
dtype: int64

In [ ]:
df2.sum(axis=1)

Labo      12063
Damas     64571
Spark     22855
Nexia     56105
Cobalt    81218
Gentra    56634
dtype: int64

In [ ]:
df.mean()

Maydoni (kv.km)    3.399231e+04
Aholisi            1.765062e+06
dtype: float64

In [ ]:
df2.mean()

2020    28118.666667
2021    20789.000000
dtype: float64

In [ ]:
df2.mean(axis=1)

Labo       6031.5
Damas     32285.5
Spark     11427.5
Nexia     28052.5
Cobalt    40609.0
Gentra    28317.0
dtype: float64

In [ ]:
df

Maydoni (kv.km)    Aholisi
Boʻlinishi                                               
Andijon viloyati                        4200.0  1899000.0
Buxoro viloyati                        39400.0  1384700.0
Fargʻona viloyati                       6800.0  2597000.0
Jizzax viloyati                        20500.0   910500.0
Xorazm viloyati                         6300.0  1200000.0
Namangan viloyati                       7900.0  1862000.0
Navoiy viloyati                       110800.0   767500.0
Qashqadaryo viloyati                   28400.0  2029000.0
Qoraqalpogʻiston Respublikasi         160000.0  1200000.0
Samarqand viloyati                     16400.0  2322000.0
Sirdaryo viloyati                       5100.0   648100.0
Surxondaryo viloyati                   20800.0  1676000.0
Toshkent viloyati                      15300.0  4450000.0

In [ ]:
df.idxmin()

Maydoni (kv.km)     Andijon viloyati
Aholisi            Sirdaryo viloyati
dtype: object

In [ ]:
df.idxmax()

Maydoni (kv.km)    Qoraqalpogʻiston Respublikasi
Aholisi                        Toshkent viloyati
dtype: object

Locating index number by columns

In [ ]:
df['Maydoni (kv.km)'].argmin()

0

In [ ]:
df['Aholisi'].argmin()

10

In [ ]:
df.iloc[0]

Maydoni (kv.km)       4200.0
Aholisi            1899000.0
Name: Andijon viloyati, dtype: float64

In [ ]:
df.iloc[10]

Maydoni (kv.km)      5100.0
Aholisi            648100.0
Name: Sirdaryo viloyati, dtype: float64

In [ ]:
df['Maydoni (kv.km)'].argmax()

8

In [ ]:
df.iloc[8]

Maydoni (kv.km)     160000.0
Aholisi            1200000.0
Name: Qoraqalpogʻiston Respublikasi, dtype: float64

In [ ]:
df['Aholisi'].argmax()

12

In [ ]:
df.iloc[12]

Maydoni (kv.km)      15300.0
Aholisi            4450000.0
Name: Toshkent viloyati, dtype: float64

Dealing with a bigger tables

In [ ]:
df=pd.read_csv('https://raw.githubusercontent.com/anvarnarz/praktikum_datasets/main/merc.csv')
df.head()

model  year  price transmission  mileage fuelType  tax   mpg  \
0        SLK  2005   5200    Automatic    63000   Petrol  325  32.1   
1    S Class  2017  34948    Automatic    27000   Hybrid   20  61.4   
2   SL CLASS  2016  49948    Automatic     6200   Petrol  555  28.0   
3    G Class  2016  61948    Automatic    16000   Petrol  325  30.4   
4    G Class  2016  73948    Automatic     4000   Petrol  325  30.1   

   engineSize  
0         1.8  
1         2.1  
2         5.5  
3         4.0  
4         4.0

In [ ]:
df.tail()

model  year  price transmission  mileage fuelType  tax   mpg  \
13114     C Class  2020  35999    Automatic      500   Diesel  145  55.4   
13115     B Class  2020  24699    Automatic     2500   Diesel  145  55.4   
13116   GLC Class  2019  30999    Automatic    11612   Diesel  145  41.5   
13117   CLS Class  2019  37990    Automatic     2426   Diesel  145  45.6   
13118     S Class  2019  54999    Automatic     2075   Diesel  145  52.3   

       engineSize  
13114         2.0  
13115         2.0  
13116         2.1  
13117         2.0  
13118         2.9

In [ ]:
df.sample(10)

model  year  price transmission  mileage fuelType  tax   mpg  \
9638    CLS Class  2016  23995    Automatic    13062   Diesel  150  51.4   
7394    CLS Class  2019  37500    Semi-Auto     2609   Diesel  145  48.7   
6567      C Class  2016  40999    Semi-Auto    19500   Petrol  300  32.8   
1423      E Class  2016  16499    Semi-Auto    40930   Diesel  150  54.3   
8315      C Class  2019  19699    Semi-Auto    15020   Petrol  145  46.3   
10608     A Class  2019  18748    Automatic    14153   Diesel  145  68.9   
5853      E Class  2020  46200    Semi-Auto     4814   Petrol  145  33.7   
4636    GLA Class  2015  17000    Semi-Auto    58000   Diesel  125  55.4   
8538      C Class  2018  28699    Semi-Auto     5234   Diesel  145  57.7   
8847      B Class  2020  24419    Semi-Auto     1500   Diesel  145  55.4   

       engineSize  
9638          3.0  
7394          2.9  
6567          4.0  
1423          3.0  
8315          1.5  
10608         1.5  
5853          2.0  
4636          2.1  
8538          2.0  
8847          2.0

There are 13119 elements in each column, meaning it is well completed table.

non-null is a number.

Dtype is data type.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13119 entries, 0 to 13118
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   model         13119 non-null  object 
 1   year          13119 non-null  int64  
 2   price         13119 non-null  int64  
 3   transmission  13119 non-null  object 
 4   mileage       13119 non-null  int64  
 5   fuelType      13119 non-null  object 
 6   tax           13119 non-null  int64  
 7   mpg           13119 non-null  float64
 8   engineSize    13119 non-null  float64
dtypes: float64(2), int64(4), object(3)
memory usage: 922.6+ KB


Representing only numbers in the table

(std) Standard deviation is a fundamental concept in statistics that measures how spread out a set of data is relative to its average (mean). It tells you how much, on average, each data point deviates from the mean.

In [ ]:
df.describe()

year          price        mileage           tax           mpg  \
count  13119.000000   13119.000000   13119.000000  13119.000000  13119.000000   
mean    2017.296288   24698.596920   21949.559037    129.972178     55.155843   
std        2.224709   11842.675542   21176.512267     65.260286     15.220082   
min     1970.000000     650.000000       1.000000      0.000000      1.100000   
25%     2016.000000   17450.000000    6097.500000    125.000000     45.600000   
50%     2018.000000   22480.000000   15189.000000    145.000000     56.500000   
75%     2019.000000   28980.000000   31779.500000    145.000000     64.200000   
max     2020.000000  159999.000000  259000.000000    580.000000    217.300000   

         engineSize  
count  13119.000000  
mean       2.071530  
std        0.572426  
min        0.000000  
25%        1.800000  
50%        2.000000  
75%        2.100000  
max        6.200000

Correlation in a pandas DataFrame refers to the statistical relationship between two or more columns within the DataFrame. It measures how much one column tends to move with or against another.

There are 2 ways to calculate correlation in pandas:

a) Using df.corr(): This method computes the pairwise correlation coefficients between all numerical columns in the DataFrame. It returns a new DataFrame with the correlation values.

b) Using df.corrwith(): This method calculates the correlation between a specific column and other columns (or another DataFrame).

(minus means reverse correlation)

In [ ]:
df.corr()

<ipython-input-6-2f6f6606aa2c>:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  df.corr()


year     price   mileage       tax       mpg  engineSize
year        1.000000  0.520712 -0.738027  0.012480 -0.094626   -0.142147
price       0.520712  1.000000 -0.537214  0.268717 -0.438445    0.516126
mileage    -0.738027 -0.537214  1.000000 -0.160223  0.202850    0.063652
tax         0.012480  0.268717 -0.160223  1.000000 -0.513742    0.338341
mpg        -0.094626 -0.438445  0.202850 -0.513742  1.000000   -0.339862
engineSize -0.142147  0.516126  0.063652  0.338341 -0.339862    1.000000

In [ ]:
df.corrwith(df['price'])

<ipython-input-10-4af89057fcc9>:1: FutureWarning: The default value of numeric_only in DataFrame.corrwith is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  df.corrwith(df['price'])


year          0.520712
price         1.000000
mileage      -0.537214
tax           0.268717
mpg          -0.438445
engineSize    0.516126
dtype: float64

In [ ]:
df['price'].corr(df['mileage'])

-0.5372141241046461

In [ ]:
df['engineSize'].corr(df['mpg'])

-0.3398624345828943

the .unique() method is used to extract the unique values from a specific column or the entire DataFrame.

There are 13118 rows in DataFrame, but let's see how many models there are.

In [ ]:
df

model  year  price transmission  mileage fuelType  tax   mpg  \
0             SLK  2005   5200    Automatic    63000   Petrol  325  32.1   
1         S Class  2017  34948    Automatic    27000   Hybrid   20  61.4   
2        SL CLASS  2016  49948    Automatic     6200   Petrol  555  28.0   
3         G Class  2016  61948    Automatic    16000   Petrol  325  30.4   
4         G Class  2016  73948    Automatic     4000   Petrol  325  30.1   
...           ...   ...    ...          ...      ...      ...  ...   ...   
13114     C Class  2020  35999    Automatic      500   Diesel  145  55.4   
13115     B Class  2020  24699    Automatic     2500   Diesel  145  55.4   
13116   GLC Class  2019  30999    Automatic    11612   Diesel  145  41.5   
13117   CLS Class  2019  37990    Automatic     2426   Diesel  145  45.6   
13118     S Class  2019  54999    Automatic     2075   Diesel  145  52.3   

       engineSize  
0             1.8  
1             2.1  
2             5.5  
3             4.0  
4             4.0  
...           ...  
13114         2.0  
13115         2.0  
13116         2.1  
13117         2.0  
13118         2.9  

[13119 rows x 9 columns]

In [ ]:
models = df['model'].unique()
models.sort()
models

array([' A Class', ' B Class', ' C Class', ' CL Class', ' CLA Class',
       ' CLC Class', ' CLK', ' CLS Class', ' E Class', ' G Class',
       ' GL Class', ' GLA Class', ' GLB Class', ' GLC Class',
       ' GLE Class', ' GLS Class', ' M Class', ' R Class', ' S Class',
       ' SL CLASS', ' SLK', ' V Class', ' X-CLASS', '180', '200', '220',
       '230'], dtype=object)

In [ ]:
fuels = df['fuelType'].unique()
fuels.sort()
fuels

array(['Diesel', 'Hybrid', 'Other', 'Petrol'], dtype=object)

In [ ]:
trans = df['transmission'].unique()
trans.sort()
trans

array(['Automatic', 'Manual', 'Other', 'Semi-Auto'], dtype=object)


In pandas DataFrames, the .value_counts() method is a powerful tool for analyzing the frequency of unique values within a specific column or across the entire DataFrame.

Functionality:

Counting unique values: It calculates the number of times each unique value appears in the specified column(s).
Output: It returns a pandas Series object, where the index represents the unique values and the values represent their corresponding counts.

Syntax:

Basic usage: df['column_name'].value_counts()
Counting in all columns: df.value_counts()

In [ ]:
df['model'].value_counts()

 C Class      3747
 A Class      2561
 E Class      1953
 GLC Class     960
 GLA Class     847
 B Class       591
 CL Class      511
 GLE Class     461
 SL CLASS      260
 CLS Class     237
 V Class       207
 S Class       197
 GL Class      121
 SLK            95
 CLA Class      86
 X-CLASS        82
 M Class        79
 GLS Class      74
 GLB Class      19
 G Class        15
 CLK             7
 CLC Class       3
 R Class         2
180              1
230              1
220              1
200              1
Name: model, dtype: int64

In [ ]:
df['fuelType'].value_counts()

Diesel    9187
Petrol    3752
Hybrid     173
Other        7
Name: fuelType, dtype: int64

In [ ]:
df['transmission'].value_counts()

Semi-Auto    6848
Automatic    4825
Manual       1444
Other           2
Name: transmission, dtype: int64

Filtering

The .isin() method is a versatile tool for checking if values in a column exist within a specified list, Series, or other iterables. It returns a DataFrame of boolean values, indicating whether each element in the column matches any value in the reference set.

In [ ]:
models = [' B Class', ' CLK']
filtr = df['model'].isin(models)
filtr

0        False
1        False
2        False
3        False
4        False
         ...  
13114    False
13115     True
13116    False
13117    False
13118    False
Name: model, Length: 13119, dtype: bool

In [ ]:
df[filtr]

model  year  price transmission  mileage fuelType  tax   mpg  \
12      B Class  2019  23750    Automatic      278   Diesel  145  55.4   
15      B Class  2017  15890    Automatic    24841   Diesel  150  68.9   
18      B Class  2017  17513    Automatic    14128   Diesel  150  68.9   
45      B Class  2019  20945    Automatic    13661   Petrol  145  45.6   
54      B Class  2010   3890    Automatic    71000   Petrol  240  40.9   
...         ...   ...    ...          ...      ...      ...  ...   ...   
13031   B Class  2019  22749    Automatic    10101   Diesel  145  60.1   
13033   B Class  2016  14599    Automatic    28156   Diesel   20  68.9   
13073   B Class  2019  19350    Automatic    13674   Petrol  145  45.6   
13111   B Class  2019  19344    Automatic    15257   Petrol  145  45.6   
13115   B Class  2020  24699    Automatic     2500   Diesel  145  55.4   

       engineSize  
12            2.0  
15            1.5  
18            2.1  
45            1.3  
54            1.5  
...           ...  
13031         1.5  
13033         2.1  
13073         1.3  
13111         1.3  
13115         2.0  

[598 rows x 9 columns]

In [ ]:
filtr = df['fuelType'].isin(['Diesel', 'Petrol'])
df[filtr]

model  year   price transmission  mileage fuelType  tax   mpg  \
0             SLK  2005    5200    Automatic    63000   Petrol  325  32.1   
2        SL CLASS  2016   49948    Automatic     6200   Petrol  555  28.0   
3         G Class  2016   61948    Automatic    16000   Petrol  325  30.4   
4         G Class  2016   73948    Automatic     4000   Petrol  325  30.1   
5        SL CLASS  2011  149948    Automatic     3000   Petrol  570  21.4   
...           ...   ...     ...          ...      ...      ...  ...   ...   
13114     C Class  2020   35999    Automatic      500   Diesel  145  55.4   
13115     B Class  2020   24699    Automatic     2500   Diesel  145  55.4   
13116   GLC Class  2019   30999    Automatic    11612   Diesel  145  41.5   
13117   CLS Class  2019   37990    Automatic     2426   Diesel  145  45.6   
13118     S Class  2019   54999    Automatic     2075   Diesel  145  52.3   

       engineSize  
0             1.8  
2             5.5  
3             4.0  
4             4.0  
5             6.2  
...           ...  
13114         2.0  
13115         2.0  
13116         2.1  
13117         2.0  
13118         2.9  

[12939 rows x 9 columns]

In [ ]:
df[df.year>2018]

model  year   price transmission  mileage fuelType  tax   mpg  \
8         G Class  2019  139948    Automatic    12000   Petrol  145  21.4   
10        A Class  2020   32980    Automatic      606   Petrol  145  35.8   
11        A Class  2020   26980    Automatic     1000   Diesel  145  62.8   
12        B Class  2019   23750    Automatic      278   Diesel  145  55.4   
13      GLC Class  2019   67980    Automatic     5500   Petrol  145  21.7   
...           ...   ...     ...          ...      ...      ...  ...   ...   
13114     C Class  2020   35999    Automatic      500   Diesel  145  55.4   
13115     B Class  2020   24699    Automatic     2500   Diesel  145  55.4   
13116   GLC Class  2019   30999    Automatic    11612   Diesel  145  41.5   
13117   CLS Class  2019   37990    Automatic     2426   Diesel  145  45.6   
13118     S Class  2019   54999    Automatic     2075   Diesel  145  52.3   

       engineSize  
8             4.0  
10            2.0  
11            1.5  
12            2.0  
13            4.0  
...           ...  
13114         2.0  
13115         2.0  
13116         2.1  
13117         2.0  
13118         2.9  

[5272 rows x 9 columns]

In [ ]:
df[(df.year>2018) & (df.price<20000)]

model  year  price transmission  mileage fuelType  tax   mpg  \
89      A Class  2019  18118    Automatic    15561   Diesel  145  68.9   
90      A Class  2019  17990    Automatic    14783   Diesel  145  65.7   
93      A Class  2019  19840    Automatic     6781   Petrol  145  28.5   
94      A Class  2019  18241    Automatic    15722   Diesel  145  65.7   
95      A Class  2019  18141    Automatic    15689   Diesel  145  65.7   
...         ...   ...    ...          ...      ...      ...  ...   ...   
12852   C Class  2019  19750    Automatic    12498   Petrol  145  46.3   
12961   A Class  2019  19888       Manual     9071   Petrol  145  53.3   
12993   B Class  2019  19345    Automatic    14082   Petrol  145  45.6   
13073   B Class  2019  19350    Automatic    13674   Petrol  145  45.6   
13111   B Class  2019  19344    Automatic    15257   Petrol  145  45.6   

       engineSize  
89            1.5  
90            1.5  
93            1.3  
94            1.5  
95            1.5  
...           ...  
12852         1.5  
12961         1.3  
12993         1.3  
13073         1.3  
13111         1.3  

[255 rows x 9 columns]

In [ ]:
df2 = df[(df.year>=2018) & (df.price<20000) & (df.mileage<10000)]
df2

model  year  price transmission  mileage fuelType  tax   mpg  \
93        A Class  2019  19840    Automatic     6781   Petrol  145  28.5   
238       A Class  2018  19995    Semi-Auto     2348   Petrol  145  53.3   
244       A Class  2019  18995    Semi-Auto     3150   Diesel  145  65.7   
420       A Class  2019  18498       Manual     2810   Petrol  145  28.5   
426       A Class  2018  19998    Semi-Auto     6340   Petrol  145  53.3   
...           ...   ...    ...          ...      ...      ...  ...   ...   
11995         220  2018  19995    Automatic     9814   Diesel  145  61.4   
12279     A Class  2019  19450       Manual     8947   Petrol  150  28.5   
12419   CLA Class  2018  19950    Automatic     6466   Petrol  150  52.3   
12961     A Class  2019  19888       Manual     9071   Petrol  145  53.3   
13013     B Class  2018  17995    Automatic     6025   Diesel  145  68.9   

       engineSize  
93            1.3  
238           1.3  
244           1.5  
420           1.3  
426           1.3  
...           ...  
11995         2.0  
12279         1.3  
12419         1.6  
12961         1.3  
13013         2.1  

[145 rows x 9 columns]

In [ ]:
df[df.transmission == 'Automatic']

model  year  price transmission  mileage fuelType  tax   mpg  \
0             SLK  2005   5200    Automatic    63000   Petrol  325  32.1   
1         S Class  2017  34948    Automatic    27000   Hybrid   20  61.4   
2        SL CLASS  2016  49948    Automatic     6200   Petrol  555  28.0   
3         G Class  2016  61948    Automatic    16000   Petrol  325  30.4   
4         G Class  2016  73948    Automatic     4000   Petrol  325  30.1   
...           ...   ...    ...          ...      ...      ...  ...   ...   
13114     C Class  2020  35999    Automatic      500   Diesel  145  55.4   
13115     B Class  2020  24699    Automatic     2500   Diesel  145  55.4   
13116   GLC Class  2019  30999    Automatic    11612   Diesel  145  41.5   
13117   CLS Class  2019  37990    Automatic     2426   Diesel  145  45.6   
13118     S Class  2019  54999    Automatic     2075   Diesel  145  52.3   

       engineSize  
0             1.8  
1             2.1  
2             5.5  
3             4.0  
4             4.0  
...           ...  
13114         2.0  
13115         2.0  
13116         2.1  
13117         2.0  
13118         2.9  

[4825 rows x 9 columns]

In [ ]:
df[(df.transmission == 'Automatic') & (df.fuelType == 'Hybrid')]

model  year  price transmission  mileage fuelType  tax    mpg  \
1       S Class  2017  34948    Automatic    27000   Hybrid   20   61.4   
77      S Class  2016  24880    Automatic    75000   Hybrid  190   42.4   
137     C Class  2017  21000    Automatic    25000   Hybrid  140  134.5   
427     C Class  2016  19698    Automatic    25107   Hybrid    0  134.5   
1150    C Class  2016  16998    Automatic    59208   Hybrid    0   76.4   
...         ...   ...    ...          ...      ...      ...  ...    ...   
12687   C Class  2017  17480    Automatic    47019   Hybrid    0  134.5   
12707   S Class  2019  56890    Automatic     2931   Hybrid  135  100.9   
12713   C Class  2017  20480    Automatic    30053   Hybrid    0  134.5   
12867   S Class  2017  36890    Automatic    22657   Hybrid   20   61.4   
12999   C Class  2016  15999    Automatic    66077   Hybrid    0  134.5   

       engineSize  
1             2.1  
77            3.5  
137           2.0  
427           2.0  
1150          2.1  
...           ...  
12687         2.0  
12707         3.0  
12713         2.0  
12867         2.1  
12999         2.0  

[104 rows x 9 columns]